cont/base/springcontent/LuaGadgets/Gadgets/control_guard.lua

```diff
@@ -0,0 +1,128 @@
+
+function gadget:GetInfo()
+	return {
+		name = "Control guard out of map",
+		desc = "Prevent units guarding other units out of map, unless both are builders",
+		license = "GNU GPL, v2 or later",
+		layer = 0,
+		enabled = true
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then
+	return false
+end
+
+local spGetUnitCurrentCommand = Spring.GetUnitCurrentCommand
+local spGetUnitPosition = Spring.GetUnitPosition
+local spGetUnitDefID = Spring.GetUnitDefID
+local spValidUnitID = Spring.ValidUnitID
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+local spIsPosInMap = Spring.IsPosInMap
+
+local CMD_GUARD = CMD.GUARD
+local CMD_INSERT = CMD.INSERT
+local CMD_INTERNAL = CMD.OPT_INTERNAL
+
+local math_bit_and = math.bit_and
+
+local MAX_TIME_OUT_OF_MAP = Game.gameSpeed*5
+
+local isMobileUnit = {}
+local isBuilder = {}
+
+for unitDefID, udef in pairs(UnitDefs) do
+	if not (udef.isBuilding or udef.speed == 0) then
+		isMobileUnit[unitDefID] = true
+	end
+	if udef.isBuilder and (udef.buildSpeed and udef.buildSpeed > 0) and (udef.buildDistance and udef.buildDistance > 0) then

```

Comment:
I'm fairly sure these can't be nil
```suggestion
	if udef.isBuilder and udef.buildSpeed > 0 and udef.buildDistance > 0 then
```
and I _think_ that if any of the two others were 0 then `isBuilder` would become `false` so maybe (but not 100% sure) it's even as simple as
```suggestion
	if udef.isBuilder then
```